pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+            files = backup_state["files"]
+            for delta_file in files:
+                snapshot_file = SnapshotFile.parse_obj(delta_file)
+                if snapshot_file.hexdigest:
+                    all_snapshot_files[snapshot_file.hexdigest] = snapshot_file
+
+        return all_snapshot_files
+
+    def _delta_upload_hexdigest(self, *, temp_dir, chunk_path, file_obj, callback_queue, relative_path):
+        start_time = time.monotonic()
+
+        input_size = file_obj.seek(0, os.SEEK_END)
+        file_obj.seek(0)
+
+        uploaded_hash = hashlib.blake2s()
+
+        def update_hash(data):
+            uploaded_hash.update(data)
+
+        with NamedTemporaryFile(dir=temp_dir, prefix=os.path.basename(chunk_path), suffix=".tmp") as raw_output_obj:
+            rohmufile.write_file(
+                input_obj=file_obj,
+                output_obj=raw_output_obj,
+                compression_algorithm=self.compression_data.algorithm,
+                compression_level=self.compression_data.level,
+                rsa_public_key=self.encryption_data.rsa_public_key,
+                log_func=self.log.info,
+                data_callback=update_hash
+            )
+            result_size = raw_output_obj.tell()
+            raw_output_obj.seek(0)
+
+            os.link(raw_output_obj.name, chunk_path)
+
+        rohmufile.log_compression_result(
+            encrypted=bool(self.encryption_data.encryption_key_id),
+            elapsed=time.monotonic() - start_time,
+            original_size=input_size,
+            result_size=result_size,
+            source_name="$PGDATA basebackup delta file",
+            log_func=self.log.info,
+        )
+
+        size_ratio = result_size / input_size
+        self.metrics.gauge(
+            "pghoard.compressed_size_ratio",
+            size_ratio,
+            tags={
+                "algorithm": self.compression_data.algorithm,
+                "site": self.site,
+                "type": "basebackup_delta",
+            }
+        )
+
+        metadata = {
+            "compression-algorithm": self.compression_data.algorithm,
+            "encryption-key-id": self.encryption_data.encryption_key_id,
+            "format": BaseBackupFormat.delta_v1,
+            "original-file-size": input_size,
+            "host": socket.gethostname(),
+        }
+
+        self.transfer_queue.put({
+            "callback_queue": callback_queue,
+            "file_size": result_size,
+            "filetype": "basebackup_delta",
+            "local_path": chunk_path,
+            "metadata": metadata,
+            "opaque": {
+                "object_path": os.path.basename(chunk_path),
+                "hexdigest": uploaded_hash.hexdigest(),
+                "relative_path": relative_path,
+                "new_size": input_size,
+                "result_size": result_size,
+            },
+            "site": self.site,
+            "type": "UPLOAD",
+        })
+
```

Comment:
This has no limiting for how much data may be pending for upload. If transfer speed is low this will very quickly consume all available disk space if temp files are stored on a smaller volume (as they are in our case). For regular backups we have `basebackup_chunks_in_progress` controlling the rough local disk usage, single chunk is 2 gigs (likely smaller when compressed) so local disk usage is never more than `basebackup_chunks_in_progress * 2 * 1024 * 1024 * 1024`. Instead of tracking disk usage the upload could be done directly from here as suggested in other comment, which could potentially simplify things overall (though in that case we'd have fewer upload threads than usual, which could slow down upload speed).

Reply:
Now the limiting is done by the uploading threads pool size.